lib/puppet/environments.rb

```diff
@@ -230,20 +233,22 @@ def create_environment(name)
       env
     end
 
-    def valid_directory?(envdir)
-      name = Puppet::FileSystem.basename_string(envdir)
-      Puppet::FileSystem.directory?(envdir) &&
-         Puppet::Node::Environment.valid_name?(name)
+    def validated_directory(envdir)
+      env_name = Puppet::FileSystem.basename_string(envdir)
+      envdir = Puppet::Environments::Directories.real_path(envdir)
+      return nil unless Puppet::FileSystem.directory?(envdir) && Puppet::Node::Environment.valid_name?(env_name)
```

Comment:
I don't know if this is just a me thing, but I much prefer a conditional the returns false/nil explicitly rather than two sequential `return`s
eg
```rb
if directory?(envdir) && valid_name?(name)
  name
else
  nil
end
```

Reply:
heh, yeah, this is something I picked up from Eric G on RE. I think the idea is that the "happy-path" case shouldn't be in a conditional?

Reply:
I think I disagree with that, though I'm not sure exactly if I follow the rationale.

I'd argue that save for a quick input sanity check at the top of a method early return should be avoided (and even those should be minimized).

I'm aware that this method is also too small to care a lot about, but now I'm curious what other's opinions are?

/cc @tvpartytonight @Magisus @pcarlisle ?

Reply:
The sentiment I've heard is to avoid large indented blocks, which goes along with what Justin described: if you need to bail before doing any work because something is wrong, an early return is good for that, instead of having large chunks wrapped under `if` and `else`. In this case, since the cases are small and clearly binary, I think I prefer Justin's. Another plus, it avoids the `unless` ;)

Reply:
What's the rationale for avoiding early `return`s?

Reply:
It can be confusing. Again this method isn't large enough to really suffer from it, but returns in the middle of a method body can be easily missed and when you have sequential returns based on different conditionals I think they can sometimes interact in non-obvious ways. In that case I think it's usually clearer to put all the conditionals in a single control structure (if/elsif/else block) to be explicit about how they interact.

Reply:
mmkay, I suppose that's fair. I updated it.